src/vm/eventpipebuffer.cpp

```diff
@@ -54,35 +54,65 @@ bool EventPipeBuffer::WriteEvent(Thread *pThread, EventPipeSession &session, Eve
         NOTHROW;
         GC_NOTRIGGER;
         MODE_ANY;
-        PRECONDITION(pThread != NULL);
         PRECONDITION(((size_t)m_pCurrent % AlignmentSize) == 0);
     }
     CONTRACTL_END;
 
     // Calculate the size of the event.
     unsigned int eventSize = sizeof(EventPipeEventInstance) + payload.GetSize();
+    DWORD osThreadId;
 
     // Make sure we have enough space to write the event.
     if(m_pCurrent + eventSize >= m_pLimit)
     {
         return false;
     }
 
+    if (pThread == NULL)
+    {
+        // For ThreadID just get the Current Thread ID
+        osThreadId = ::GetCurrentThreadId();
+    }
+    else
+    {
+        osThreadId = pThread->GetOSThreadId();
+    }
+
     // Calculate the location of the data payload.
     BYTE *pDataDest = m_pCurrent + sizeof(EventPipeEventInstance);
+    EventPipeEventInstance *pInstance;
 
     bool success = true;
     EX_TRY
     {
+        LPCGUID pThreadActivityID = pActivityId;
+
         // Placement-new the EventPipeEventInstance.
-        EventPipeEventInstance *pInstance = new (m_pCurrent) EventPipeEventInstance(
-            session,
-            event,
-            pThread->GetOSThreadId(),
-            pDataDest,
-            payload.GetSize(),
-            pActivityId,
-            pRelatedActivityId);
+
+        if (pThread == NULL)
+        {
+            // if pthread is NULL, it's likely we are running in something like a GC thread which is not a Thread object, so it can't have an activity ID set anyway
+            pInstance = new (m_pCurrent) EventPipeEventInstance(
```

Comment:
The only thing thing different between the two cases is the activity ID argument. You can just have the condition for that one argument without duplicating the whole thing.

`(pThread != NULL) ? pActivityId : NULL`